### PR TITLE
keycloak_group: fix subgroup creation in Keycloak ≥23

### PR DESCRIPTION
**Pull request by vgaudard**
##### SUMMARY

In Keycloak versions 23 and later, GroupRepresentation.subGroups has been replaced by GroupRepresentation.subGroupCount and another endpoint to get subgroups.
This caused community.general.keycloak_group to fail when creating a group with a parent.
We now execute this second request if needed.

Fixes ansible-collections/community.general#8788

See https://www.keycloak.org/docs/latest/upgrading/#grouprepresentation-changes

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

keycloak_group

##### ADDITIONAL INFORMATION

Related issue : ansible-collections/community.general#8788

Tested on Keycloak versions (Docker images) :
- 17.0.0
- 21.1.2
- 22.0.5
- 23.0.4
- 23.0.7
- 24.0.5
- 25.0.6

### CODE REVIEW

**Comment by vgaudard** [comment]
I wasn't able to run tests locally (either before or after my changes). Because they're marked `unsupported`, I assumed this is expected.

```
$ ansible-test integration keycloak_group --docker -v
Configured locale: C.UTF-8
Falling back to tests in "tests/integration/targets/" because "roles/test/" was not found.
WARNING: Excluding target tests marked "unsupported" which require --allow-unsupported or prefixing with "unsupported/": keycloak_group
WARNING: All targets skipped.

$ ansible-test integration keycloak_group --docker -v --allow-unsupported
Configured locale: C.UTF-8
Falling back to tests in "tests/integration/targets/" because "roles/test/" was not found.
Run command: docker -v
Detected "docker" container runtime version: Docker version 26.1.4, build 5650f9b
Run command: docker info --format '{{ json . }}'
Run command: docker version --format '{{ json . }}'
Container runtime: docker client=26.1.4 server=26.1.4 cgroup=v1 DD+WSL2
Run command: docker image inspect quay.io/ansible/ansible-test-utility-container:2.0.0
Run command: docker run --volume /sys/fs/cgroup:/probe:ro --name ansible-test-probe-yPUyPpLc --rm quay.io/ansible/ansible-test-utility-container:2.0.0 sh -c 'audit-status && cat /proc ...
Container host audit status: EPERM (-1)
Container host max open files: 1048576
Container loginuid: 4294967295 (not set)
Assuming Docker is available on localhost.
Run command with data: docker run --tmpfs /tmp:exec --tmpfs /run:exec --tmpfs /run/lock --volume /var/run/docker.sock:/var/run/docker.sock --cgroupns host --tmpfs /sys/fs/cgroup --vol ...
ERROR: Host DockerConfig(python=NativePythonConfig(version='3.12', path='/usr/bin/python3.12'), name='default', image='quay.io/ansible/default-test-container:8.12.0', memory=None, privileged=False, seccomp='default', cgroup=CGroupVersion.V1_V2, audit=AuditMode.REQUIRED) job failed:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/provisioning.py", line 200, in dispatch_jobs
    thread.wait_for_result()
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/thread.py", line 44, in wait_for_result
    raise exception[1].with_traceback(exception[2])
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/thread.py", line 34, in run
    self._result.put((self.action(), None))
                      ^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/provisioning.py", line 134, in provision
    profile.provision()
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/host_profiles.py", line 456, in provision
    container = run_support_container(
                ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/containers.py", line 135, in run_support_container
    current_container_id = get_docker_container_id()
                           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/thread.py", line 59, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/util.py", line 151, in cache_func
    value = storage[None] = func()
                            ^^^^^^
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/docker_util.py", line 599, in get_docker_container_id
    mounts = MountEntry.loads(mountinfo_path.read_text())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/cgroup.py", line 107, in loads
    return tuple(cls.parse(line) for line in value.splitlines())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/cgroup.py", line 107, in <genexpr>
    return tuple(cls.parse(line) for line in value.splitlines())
                 ^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/ansible_test/_internal/cgroup.py", line 86, in parse
    assert separator == '-'
           ^^^^^^^^^^^^^^^^
AssertionError:
FATAL: Host job(s) failed. See previous error(s) for details.
```

File: changelogs/fragments/8979-keycloak_group-fix-subgroups.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - keycloak_group - fix crash caused in subgroup creation. The crash was caused by a missing or empty `subGroups` property in Keycloak ≥23
```
Comment (felixfontein) [inline changelogs/fragments/8979-keycloak_group-fix-subgroups.yml:2]:
```suggestion
  - keycloak_group - fix crash caused in subgroup creation. The crash was caused by a missing or empty ``subGroups`` property in Keycloak ≥23 (https://github.com/ansible-collections/community.general/issues/8788, https://github.com/ansible-collections/community.general/pull/8979).
```

**Comment by vgaudard** [comment]
@UKFr-DIZ This is a fix for issue #8788 .

@vvanouytsel This PR might fix issue #8366 , could you confirm that the issue happened with Keycloak version 23?
Details : I suspect this was caused by Keycloak returning a `subGroupCount > 0` and `subGroups == []` after creating the subgroup. This caused `after_group` to be set to `None`, and so the module failed when formatting the result message.

This pull request is ready for review.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added a first comment below.

Regarding the tests: `unsupported` means they don't run in CI, you have to run them manually. (See also `tests/integration/targets/keycloak_group/readme.adoc` in case you haven't seen it yet.) The error you are seeing seems to be a bug in ansible-core, or a problem with Docker on your machine, or a combination of your Docker and a bug in ansible-core, and are unrelated to the tests for this module.

(Did you try passing `--python 3.12` explicitly? Usually if you don't provide a OS docker image like `--docker ubuntu2404`, you have to specify the Python version. It could be that ansible-test picks a default Python version for the `default `image, but :shrug:)

File: changelogs/fragments/8979-keycloak_group-fix-subgroups.yml (hunk 1 above)
Comment (vgaudard) [reply changelogs/fragments/8979-keycloak_group-fix-subgroups.yml:2]: Thank for your quick comment.

> Regarding the tests: unsupported means they don't run in CI, you have to run them manually.

Thanks for the explanation

> See also tests/integration/targets/keycloak_group/readme.adoc

I *may* have forgotten to read this. I have read it now.

> Did you try passing --python 3.12 explicitly?

That did the trick, kind of. I think my issue is that I'm on WSL. I also had to remove the dependency `setup_docker` to run the tests locally.

<details>
<summary>`ansible-test integration keycloak_group --allow-unsupported  --python 3.12 -vvv` [normal dependency] [truncated]</summary>

```
TASK [setup_docker : Install docker] *******************************************
task path: /home/gaudardv/ansible_collections/community/general/tests/output/.tmp/integration/keycloak_group-2rqbftlf-ÅÑŚÌβŁÈ/tests/integration/targets/setup_docker/tasks/default.yml:10
Using module file /tmp/ansible-test-ujtpqbcs/ansible/modules/apt.py
Pipelining is enabled.
<testhost> ESTABLISH LOCAL CONNECTION FOR USER: gaudardv
<testhost> EXEC /bin/sh -c 'sudo -H -S -n  -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-nvhzecpkvqoanzhdevfiukxgizhqrlyg ; /usr/bin/python3'"'"' && sleep 0'
failed: [testhost] (item=docker.io) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": "docker.io",
    "module_stderr": "sudo: a password is required\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```
</details>

Without this dependency, I was able to run the tests before and after my changes, and both passed.

File: changelogs/fragments/8979-keycloak_group-fix-subgroups.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - keycloak_group - fix crash caused in subgroup creation. The crash was caused by a missing or empty ``subGroups`` property in Keycloak ≥23 (https://github.com/ansible-collections/community.general/issues/8788, https://github.com/ansible-collections/community.general/pull/8979).
```
Comment (russoz) [inline changelogs/fragments/8979-keycloak_group-fix-subgroups.yml:2]: @felixfontein will the changelog builder handle special unicode characters? If not, then we might need to change the `≥` to `>=`.

**Review by russoz — COMMENTED** [review]
hi @vgaudard Thanks for the contribution!

Other than my question in the changelog frag, the rest LGTM.

File: changelogs/fragments/8979-keycloak_group-fix-subgroups.yml (hunk 1 above)
Comment (russoz) [reply changelogs/fragments/8979-keycloak_group-fix-subgroups.yml]: Hi @vgaudard 

You might want give it a try to `andebox` by yours truly:

https://pypi.org/project/andebox/

Tests that depend on `setup_docker` might be easier to execute using an accompanying VM. I have no idea if that would work in WSL, but locally what I would do is:

```
$ andebox vagrant -n ubuntu-jammy -s -- unsupported/keycloak_group --color yes -vvv
```

File: changelogs/fragments/8979-keycloak_group-fix-subgroups.yml (hunk 2 above)
Comment (felixfontein) [reply changelogs/fragments/8979-keycloak_group-fix-subgroups.yml:2]: The changelog generator doesn't care (as long as everything is UTF-8 encoded), and both RST and MarkDown should handle Unicode fine.

**Comment by felixfontein** [comment]
If nobody objects, I'll merge this in a week.

CC @fgruenbauer who lately worked a lot on keycloak_* modules.

**Comment by vgaudard** [comment]
Thank you both for your help.

> You might want give it a try to andebox by yours truly:

I did, but I can't setup vagrant on my work computer, so I returned to classic ansible-test and fidgeting with dependencies.

> I was able to run the tests before and after my changes, and both passed.

I ran the tests again but on Keycloak 25, and... they failed. I forgot to change another method where we want subgroups of a group (`get_subgroup_by_chain`), which caused sub-sub-groups to fail.
It should be OK now.

**Comment by russoz** [comment]
hi @vgaudard "it should be OK" is not very reassuring :-) Could you please confirm if you got the test to run successfully with kc 25? TIA

**Comment by felixfontein** [comment]
@vgaudard ping

**Comment by vgaudard** [comment]
Hi
Yes, I ran the tests successfully on both Keycloak 20 and 25.

**Comment by felixfontein** [comment]
Thanks for the confirmation!

@vgaudard thanks for your contribution!
@russoz thanks for reviewing!

**Comment by felixfontein** [comment]
FYI @vgaudard you might be interested in https://forum.ansible.com/t/keycloak-modules-in-community-general/41746

**Pull request merged**